accepted/2020/platform-checks/platform-checks.md

```diff
@@ -0,0 +1,450 @@
+                    AttributeTargets.Event |
+                    AttributeTargets.Method |
+                    AttributeTargets.Module |
+                    AttributeTargets.Property |
+                    AttributeTargets.Struct,
+                    AllowMultiple=false, Inherited=true)]
+    public sealed class AddedInOSPlatformAttribute : OSPlatformAttribute
+    {
+        public PlatformSpecificAttribute(string osPlatform,
+                                         int major,
+                                         int minor,
+                                         int build,
+                                         int revision);
+    }
+
+    [AttributeUsage(AttributeTargets.Assembly |
+                    AttributeTargets.Class |
+                    AttributeTargets.Constructor |
+                    AttributeTargets.Event |
+                    AttributeTargets.Method |
+                    AttributeTargets.Module |
+                    AttributeTargets.Property |
+                    AttributeTargets.Struct,
+                    AllowMultiple=false, Inherited=true)]
+    public sealed class RemovedInOSPlatformAttribute : OSPlatformAttribute
+    {
+        public PlatformSpecificAttribute(string osPlatform,
+                                         int major,
+                                         int minor,
+                                         int build,
+                                         int revision);
+    }
+
+    [AttributeUsage(AttributeTargets.Assembly |
+                    AttributeTargets.Class |
+                    AttributeTargets.Constructor |
+                    AttributeTargets.Event |
+                    AttributeTargets.Method |
+                    AttributeTargets.Module |
+                    AttributeTargets.Property |
+                    AttributeTargets.Struct,
+                    AllowMultiple=false, Inherited=true)]
+    public sealed class ObsoletedInOSPlatformAttribute : OSPlatformAttribute
+    {
+        public PlatformSpecificAttribute(string osPlatform,
+                                         int major,
+                                         int minor,
+                                         int build,
+                                         int revision);
+        public string Url { get; set; }
+    }
+}
+```
+
+### Platform context
+
+To determine the platform context of the call site the analyzer must consider
+the following two configurations:
+
+1. **Call-site**. Application of `AddedInOSPlatformAttribute` to the containing
+   member, type, module, or assembly.
+
+2. **Project** The project TFMs and `TargetPlatformMinVersion`. If the project
+   has an platform-specific TFM (such as `net5.0-ios13.0`) we consider it to be
+   on the specified platform. If `TargetPlatformMinVersion` is set, we assume
+   that version, otherwise we assume the version as specified by the TFM.
+
+If (2) is platform neutral (e.g. `net5.0`), the platform context is entirely
```

Comment:
If a library is platform neutral, e.g. `net5.0`, are we going to flag api's which are unavailable on some of the platforms so the code author knows their code won't work everywhere?

Reply:
Yes. For example, today the registry is available as NuGet package for .NET Standard 2.0. I would expect us to add an attribute to it that marks the entire assembly as Windows-specific. Consuming that from .NET 5 will give folks a warning unless they mark their APIs Windows specific or target `net5.0-windows`.